File: src/main/java/com/milaboratory/mixcr/trees/MutationsUtils.java
package com.milaboratory.mixcr.trees;

import com.milaboratory.core.Range;
import com.milaboratory.core.alignment.AffineGapAlignmentScoring;
import com.milaboratory.core.alignment.AlignmentScoring;
import com.milaboratory.core.mutations.Mutation;
import com.milaboratory.core.mutations.Mutations;
import com.milaboratory.core.mutations.MutationsBuilder;
import com.milaboratory.core.sequence.NucleotideAlphabet;
import com.milaboratory.core.sequence.NucleotideSequence;
import com.milaboratory.core.sequence.Wildcard;
import com.milaboratory.mixcr.util.RangeInfo;

import java.util.*;
import java.util.function.BiFunction;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

final class MutationsUtils {
    private MutationsUtils() {
    }

    /**
     * Mutate and get result by range in original sequence
     */
    static NucleotideSequence buildSequence(NucleotideSequence sequence1, Mutations<NucleotideSequence> mutations, RangeInfo rangeInfo) {
        return mutations.mutate(sequence1).getRange(projectRange(mutations, rangeInfo));
    }

    static Range projectRange(Mutations<NucleotideSequence> mutations, RangeInfo rangeInfo) {
        //for including inclusions before position one must step left before conversion and step right after
        int from = positionIfNucleotideWasDeleted(mutations.convertToSeq2Position(rangeInfo.getRange().getLower()));
        if (rangeInfo.isIncludeFirstInserts()) {
            from -= IntStream.of(mutations.getRAWMutations())
                    .filter(mutation -> Mutation.getPosition(mutation) == rangeInfo.getRange().getLower() && Mutation.isInsertion(mutation))
                    .count();
        }

        int to = positionIfNucleotideWasDeleted(mutations.convertToSeq2Position(rangeInfo.getRange().getUpper()));
        return new Range(from, to);
    }

    static MutationsDescription mutationsBetween(MutationsDescription first, MutationsDescription second) {
        return new MutationsDescription(
                mutationsBetween(first.getVMutationsWithoutCDR3(), second.getVMutationsWithoutCDR3()),
                first.getVMutationsInCDR3WithoutNDN().differenceWith(second.getVMutationsInCDR3WithoutNDN()),
                first.getKnownNDN().differenceWith(second.getKnownNDN()),
                first.getJMutationsInCDR3WithoutNDN().differenceWith(second.getJMutationsInCDR3WithoutNDN()),
                mutationsBetween(first.getJMutationsWithoutCDR3(), second.getJMutationsWithoutCDR3())
        );
    }

    private static List<MutationsWithRange> mutationsBetween(List<MutationsWithRange> firstMutations, List<MutationsWithRange> secondMutations) {
        return fold(firstMutations, secondMutations, MutationsWithRange::differenceWith);
    }

    static MutationsWithRange intersection(MutationsWithRange base, MutationsWithRange comparison, RangeInfo intersection) {
        return new MutationsWithRange(
                base.getSequence1(),
                MutationsUtils.intersection(
                        base.getMutations(),
                        comparison.getMutations(),
                        intersection
                ),
                intersection
        );
    }

    static MutationsWithRange intersection(MutationsWithRange from, MutationsWithRange to) {
        if (!from.getRangeInfo().equals(to.getRangeInfo())) {
            throw new IllegalArgumentException();
        }
        return new MutationsWithRange(
                from.getSequence1(),
                MutationsUtils.intersection(
                        from.getMutations(),
                        to.getMutations(),
                        from.getRangeInfo()
                ),
                from.getRangeInfo()
        );
    }

    static List<MutationsWithRange> intersection(List<MutationsWithRange> from, List<MutationsWithRange> to) {
        return fold(from, to, (a, b) -> intersection(a, b, a.getRangeInfo().intersection(b.getRangeInfo())));
    }

    static <T> List<T> fold(
            List<MutationsWithRange> firstMutations,
            List<MutationsWithRange> secondMutations,
            BiFunction<MutationsWithRange, MutationsWithRange, T> folder
    ) {
        if (firstMutations.size() != secondMutations.size()) {
            throw new IllegalArgumentException();
        }
        return IntStream.range(0, firstMutations.size())
                .mapToObj(i -> folder.apply(firstMutations.get(0), secondMutations.get(0)))
                .collect(Collectors.toList());
    }

    static int positionIfNucleotideWasDeleted(int position) {
        if (position < -1) {
            return Math.abs(position + 1);
        }
        if (position == -1) {
            return 0;
        }
        return position;
    }

    static AlignmentScoring<NucleotideSequence> NDNScoring() {
        return new AffineGapAlignmentScoring<>(
                NucleotideSequence.ALPHABET,
                calculateSubstitutionMatrix(5, -4, 4, NucleotideSequence.ALPHABET),
                -10,
                -1
        );
    }

    private static int[] calculateSubstitutionMatrix(int match, int mismatch, int multiplierOfAsymmetry, NucleotideAlphabet alphabet) {
        int codes = alphabet.size();
        int[] matrix = new int[codes * codes];
        Arrays.fill(matrix, mismatch);
        for (int i = 0; i < codes; ++i)
            matrix[i + codes * i] = match;
        return fillWildcardScoresMatches(matrix, alphabet, match, mismatch, multiplierOfAsymmetry);
    }

    private static int[] fillWildcardScoresMatches(int[] matrix, NucleotideAlphabet alphabet, int match, int mismatch, int multiplierOfAsymmetry) {
        int alSize = alphabet.size();

        if (matrix.length != alSize * alSize)
            throw new IllegalArgumentException("Wrong matrix size.");

        //TODO remove excludeSet from milib
        for (Wildcard wc1 : alphabet.getAllWildcards())
            for (Wildcard wc2 : alphabet.getAllWildcards()) {
                if (wc1.isBasic() && wc2.isBasic())
                    continue;
                int sumScore = 0;
                for (int i = 0; i < wc1.basicSize(); i++) {
                    if (wc2.matches(wc1.getMatchingCode(i))) {
                        sumScore += match;
                    } else {
                        sumScore += mismatch;
                    }
                }
                for (int i = 0; i < wc2.basicSize(); i++) {
                    if (wc1.matches(wc2.getMatchingCode(i))) {
                        sumScore += match * multiplierOfAsymmetry;
                    } else {
                        sumScore += mismatch * multiplierOfAsymmetry;
                    }
                }
                sumScore /= wc1.basicSize() + wc2.basicSize() * multiplierOfAsymmetry;
                matrix[wc1.getCode() + wc2.getCode() * alSize] = sumScore;
            }

        return matrix;
    }

    static Mutations<NucleotideSequence> intersection(
            Mutations<NucleotideSequence> first,
            Mutations<NucleotideSequence> second,
            RangeInfo rangeInfo
    ) {
        return simpleIntersection(
                first,
                second,
                rangeInfo.getRange()
        );
    }

    //TODO removals and inserts
    private static Mutations<NucleotideSequence> simpleIntersection(
            Mutations<NucleotideSequence> first,
            Mutations<NucleotideSequence> second,
            Range range
    ) {
        Set<Integer> mutationsOfFirstAsSet = Arrays.stream(first.getRAWMutations()).boxed().collect(Collectors.toSet());

        MutationsBuilder<NucleotideSequence> mutationsBuilder = new MutationsBuilder<>(NucleotideSequence.ALPHABET);
        for (int i = 0; i < second.size(); i++) {
            int mutation = second.getMutation(i);
            int position = Mutation.getPosition(mutation);
            if (range.contains(position)) {
                if (mutationsOfFirstAsSet.contains(mutation)) {
                    mutationsBuilder.append(mutation);
                }
            }
        }
        return mutationsBuilder.createAndDestroy();
    }

    //TODO removals and inserts
    static Mutations<NucleotideSequence> concreteNDNChild(Mutations<NucleotideSequence> parent, Mutations<NucleotideSequence> child) {
        Map<Integer, Set<Integer>> mutationsOfParentByPositions = Arrays.stream(parent.getRAWMutations())
                .boxed()
                .collect(Collectors.groupingBy(Mutation::getPosition, Collectors.toSet()));
        MutationsBuilder<NucleotideSequence> mutationsBuilder = new MutationsBuilder<>(NucleotideSequence.ALPHABET);
        for (int i = 0; i < child.size(); i++) {
            int mutationOfChild = child.getMutation(i);
            if (Mutation.isInDel(mutationOfChild)) {
                mutationsBuilder.append(mutationOfChild);
            } else {
                int position = Mutation.getPosition(mutationOfChild);
                Optional<Integer> mutationsOfParent = mutationsOfParentByPositions.getOrDefault(position, Collections.emptySet()).stream()
                        .filter(Mutation::isSubstitution)
                        .findFirst();
                if (!mutationsOfParent.isPresent()) {
                    mutationsBuilder.append(mutationOfChild);
                } else {
                    byte from = Mutation.getFrom(mutationOfChild);
                    byte to = concreteChild(Mutation.getTo(mutationsOfParent.get()), Mutation.getTo(mutationOfChild));
                    if (from != to) {
                        mutationsBuilder.append(Mutation.createSubstitution(position, from, to));
                    }
                }
            }
        }

        return mutationsBuilder.createAndDestroy();
    }

    private static byte concreteChild(byte parentSymbol, byte childSymbol) {
        if (parentSymbol == childSymbol) {
            return childSymbol;
        } else if (NucleotideSequence.ALPHABET.isWildcard(childSymbol)) {
            if (matchesStrictly(NucleotideSequence.ALPHABET.codeToWildcard(childSymbol), parentSymbol)) {
                return parentSymbol;
            } else {
                long basicMask = NucleotideSequence.ALPHABET.codeToWildcard(parentSymbol).getBasicMask()
                        | NucleotideSequence.ALPHABET.codeToWildcard(childSymbol).getBasicMask();
                return NucleotideSequence.ALPHABET.maskToWildcard(basicMask).getCode();
            }
        } else {
            return childSymbol;
        }
    }

    //TODO removals and inserts
    static Mutations<NucleotideSequence> findNDNCommonAncestor(Mutations<NucleotideSequence> first, Mutations<NucleotideSequence> second) {
        Map<Integer, Set<Integer>> mutationsOfFirstByPositions = Arrays.stream(first.getRAWMutations())
                .boxed()
                .collect(Collectors.groupingBy(Mutation::getPosition, Collectors.toSet()));

        MutationsBuilder<NucleotideSequence> mutationsBuilder = new MutationsBuilder<>(NucleotideSequence.ALPHABET);
        for (int i = 0; i < second.size(); i++) {
            int mutationOfSecond = second.getMutation(i);
            int position = Mutation.getPosition(mutationOfSecond);
            Set<Integer> mutationsOfFirst = mutationsOfFirstByPositions.getOrDefault(position, Collections.emptySet());
            if (mutationsOfFirst.contains(mutationOfSecond)) {
                mutationsBuilder.append(mutationOfSecond);
            } else if (Mutation.isSubstitution(mutationOfSecond)) {
                mutationsOfFirst.stream()
                        .filter(Mutation::isSubstitution)
                        .findFirst()
                        .map(otherSubstitution -> Mutation.createSubstitution(
                                position,
                                Mutation.getFrom(mutationOfSecond),
                                combine(Mutation.getTo(mutationOfSecond), Mutation.getTo(otherSubstitution))
                        ))
                        .ifPresent(mutationsBuilder::append);
            }
        }
        return mutationsBuilder.createAndDestroy();
    }

    private static byte combine(byte firstSymbol, byte secondSymbol) {
        if (firstSymbol == secondSymbol) {
            return firstSymbol;
        } else if (NucleotideSequence.ALPHABET.isWildcard(firstSymbol) && matchesStrictly(NucleotideSequence.ALPHABET.codeToWildcard(firstSymbol), secondSymbol)) {
            return secondSymbol;
        } else if (NucleotideSequence.ALPHABET.isWildcard(secondSymbol) && matchesStrictly(NucleotideSequence.ALPHABET.codeToWildcard(secondSymbol), firstSymbol)) {
            return firstSymbol;
        } else {
            long basicMask = NucleotideSequence.ALPHABET.codeToWildcard(firstSymbol).getBasicMask()
                    | NucleotideSequence.ALPHABET.codeToWildcard(secondSymbol).getBasicMask();
            return NucleotideSequence.ALPHABET.maskToWildcard(basicMask).getCode();
        }
    }

    private static boolean matchesStrictly(Wildcard wildcard, byte secondSymbol) {
        if (!NucleotideSequence.ALPHABET.isWildcard(secondSymbol)) {
            return wildcard.matches(secondSymbol);
        } else {
            Wildcard secondAsWildcard = NucleotideSequence.ALPHABET.codeToWildcard(secondSymbol);
            return ((wildcard.getBasicMask() ^ secondAsWildcard.getBasicMask()) & secondAsWildcard.getBasicMask()) == 0;
        }
    }

    static List<MutationsWithRange> combine(List<MutationsWithRange> base, List<MutationsWithRange> combineWith) {
        return fold(base, combineWith, MutationsWithRange::combineWith);
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/trees/MutationsUtils.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
@file:Suppress("FunctionName")

package com.milaboratory.mixcr.trees

import com.milaboratory.core.Range
import com.milaboratory.core.alignment.AffineGapAlignmentScoring
import com.milaboratory.core.alignment.AlignmentScoring
import com.milaboratory.core.mutations.Mutation
import com.milaboratory.core.mutations.Mutations
import com.milaboratory.core.mutations.MutationsBuilder
import com.milaboratory.core.sequence.NucleotideAlphabet
import com.milaboratory.core.sequence.NucleotideSequence
import com.milaboratory.core.sequence.NucleotideSequence.ALPHABET
import com.milaboratory.core.sequence.Sequence
import com.milaboratory.core.sequence.Wildcard
import com.milaboratory.mixcr.util.asSequence
import com.milaboratory.mixcr.util.intersection
import io.repseq.core.GeneFeature
import java.util.*
import kotlin.math.abs

internal object MutationsUtils {
    /**
     * Mutate and get result by range in original sequence
     */
    fun <S : Sequence<S>> buildSequence(
        sequence1: S,
        mutations: Mutations<S>,
        range: Range
    ): S = mutations.mutate(sequence1).getRange(projectRange(mutations, range))

    fun <S : Sequence<S>> projectRange(mutations: Mutations<S>, range: Range): Range {
        //for including inclusions before position one must step left before conversion and step right after
        val from = positionIfNucleotideWasDeleted(mutations.convertToSeq2Position(range.lower)) -
                mutations.asSequence()
                    .count { mutation -> Mutation.getPosition(mutation) == range.lower && Mutation.isInsertion(mutation) }
        val to = positionIfNucleotideWasDeleted(mutations.convertToSeq2Position(range.upper))
        return Range(from, to)
    }

    fun difference(
        sequence1: NucleotideSequence,
        base: Mutations<NucleotideSequence>,
        comparison: Mutations<NucleotideSequence>,
        range: Range = Range(0, sequence1.size())
    ): CompositeMutations = CompositeMutations(
        sequence1,
        base,
        range,
        base.invert().combineWith(comparison)
    )

    inline fun <R, V1, V2> zip(
        first: Map<GeneFeature, V1>,
        second: Map<GeneFeature, V2>,
        function: (V1, V2, GeneFeature) -> R
    ): Map<GeneFeature, R> {
        require(first.keys == second.keys)
        return first.mapValues { (geneFeature, value) -> function(value, second[geneFeature]!!, geneFeature) }
    }

    inline fun <R, V1, V2> fold(
        first: Map<GeneFeature, V1>,
        second: Map<GeneFeature, V2>,
        function: (V1, V2, GeneFeature) -> R,
    ): Collection<R> {
        require(first.keys == second.keys)
        return first.map { (geneFeature, value) -> function(value, second[geneFeature]!!, geneFeature) }
    }

    fun positionIfNucleotideWasDeleted(position: Int): Int = when {
        position < -1 -> abs(position + 1)
        (position == -1) -> 0
        else -> position
    }

    fun NDNScoring(): AlignmentScoring<NucleotideSequence> = AffineGapAlignmentScoring(
        ALPHABET,
        calculateSubstitutionMatrix(5, -4, 4, ALPHABET),
        -10,
        -1
    )

    @Suppress("SameParameterValue")
    private fun calculateSubstitutionMatrix(
        match: Int,
        mismatch: Int,
        multiplierOfAsymmetry: Int,
        alphabet: NucleotideAlphabet
    ): IntArray {
        val codes = alphabet.size()
        val matrix = IntArray(codes * codes)
        Arrays.fill(matrix, mismatch)
        for (i in 0 until codes) matrix[i + codes * i] = match
        return fillWildcardScoresMatches(matrix, alphabet, match, mismatch, multiplierOfAsymmetry)
    }

    private fun fillWildcardScoresMatches(
        matrix: IntArray,
        alphabet: NucleotideAlphabet,
        match: Int,
        mismatch: Int,
        multiplierOfAsymmetry: Int
    ): IntArray {
        val alSize = alphabet.size()
        require(matrix.size == alSize * alSize) { "Wrong matrix size." }

        //TODO move to main library
        //TODO add comments
        //TODO remove excludeSet from milib
        for (wc1 in alphabet.allWildcards) {
            for (wc2 in alphabet.allWildcards) {
                if (wc1.isBasic && wc2.isBasic) continue
                var sumScore = 0
                for (i in 0 until wc1.basicSize()) {
                    sumScore += if (wc2.matches(wc1.getMatchingCode(i))) {
                        match
                    } else {
                        mismatch
                    }
                }
                for (i in 0 until wc2.basicSize()) {
                    sumScore += if (wc1.matches(wc2.getMatchingCode(i))) {
                        match * multiplierOfAsymmetry
                    } else {
                        mismatch * multiplierOfAsymmetry
                    }
                }
                sumScore /= wc1.basicSize() + wc2.basicSize() * multiplierOfAsymmetry
                matrix[wc1.code + wc2.code * alSize] = sumScore
            }
        }
        return matrix
    }

    /**
     * If parent was replaced we may recalculate NDN of the child.
     *
     * If there is a letter/wildcard in the child include the parent letter, replace child letter with parent letter.
     * In this case there will be no mutation from parent to child by choosing one of already possible variances of child.
     *
     * If a child letter is a wildcard, but it's not included in parent, then we need to add possibility that there
     * is no mutation between parent and child, so we replace child letter with union.
     *
     * If a child letter and the parent letter are not equal and are not wildcards, then assume that there is a mutation.
     *
     */
    //TODO removals and inserts
    fun concreteNDNChild(
        parent: Mutations<NucleotideSequence>,
        child: Mutations<NucleotideSequence>
    ): Mutations<NucleotideSequence> {
        val mutationsOfParentByPositions = parent.asSequence().groupBy { Mutation.getPosition(it) }
        val mutationsBuilder = MutationsBuilder(ALPHABET)
        child.asSequence().forEach { mutationOfChild ->
            if (Mutation.isInDel(mutationOfChild)) {
                mutationsBuilder.append(mutationOfChild)
            } else {
                val position = Mutation.getPosition(mutationOfChild)
                val mutationsOfParent =
                    mutationsOfParentByPositions[position]?.firstOrNull { Mutation.isSubstitution(it) }
                if (mutationsOfParent == null) {
                    mutationsBuilder.append(mutationOfChild)
                } else {
                    val from = Mutation.getFrom(mutationOfChild)
                    val to = concreteChild(Mutation.getTo(mutationsOfParent), Mutation.getTo(mutationOfChild))
                    if (from != to) {
                        mutationsBuilder.append(Mutation.createSubstitution(position, from.toInt(), to.toInt()))
                    }
                }
            }
        }
        return mutationsBuilder.createAndDestroy()
    }

    private fun concreteChild(parentSymbol: Byte, childSymbol: Byte): Byte = when {
        parentSymbol == childSymbol -> childSymbol
        ALPHABET.isWildcard(childSymbol) -> when {
            ALPHABET.codeToWildcard(childSymbol).matchesStrictly(parentSymbol) -> parentSymbol
            else -> ALPHABET.codeToWildcard(parentSymbol).union(ALPHABET.codeToWildcard(childSymbol))
        }
        else -> childSymbol
    }

    /**
     * Calculate post possible NDN of common ancestor. If there is uncertainty - left wildcard.
     * Input sequences may contain wildcards. This wildcards may be resolved to letter or narrower wildcard.
     *
     * If in position are different letters or not overlapping wildcards than resolve to wider wildcard.
     *
     * If in position are letter and wildcard containing this letter than resolve to this letter.
     * It's because most likely mutation occurred only in one branch.
     *
     * If in position are overlapping wildcards than resolve to the overlapping wildcard.
     * It's because most likely mutation occurred only in one branch.
     */
    //TODO removals and inserts
    fun findNDNCommonAncestor(
        first: Mutations<NucleotideSequence>,
        second: Mutations<NucleotideSequence>
    ): Mutations<NucleotideSequence> {
        val mutationsOfFirstByPositions = first.asSequence().groupBy { code -> Mutation.getPosition(code) }
        val mutationsBuilder = MutationsBuilder(ALPHABET)
        second.asSequence().forEach { mutationOfSecond ->
            val position = Mutation.getPosition(mutationOfSecond)
            val mutationsOfFirst = mutationsOfFirstByPositions[position] ?: emptyList()
            if (mutationsOfFirst.contains(mutationOfSecond)) {
                mutationsBuilder.append(mutationOfSecond)
            } else if (Mutation.isSubstitution(mutationOfSecond)) {
                val otherSubstitution = mutationsOfFirst.firstOrNull { code -> Mutation.isSubstitution(code) }
                if (otherSubstitution != null) {
                    val mutation = Mutation.createSubstitution(
                        position,
                        Mutation.getFrom(mutationOfSecond).toInt(),
                        combine(Mutation.getTo(mutationOfSecond), Mutation.getTo(otherSubstitution)).toInt()
                    )
                    mutationsBuilder.append(mutation)
                }
            }
        }
        return mutationsBuilder.createAndDestroy()
    }

    private fun combine(firstSymbol: Byte, secondSymbol: Byte): Byte = when {
        firstSymbol == secondSymbol -> firstSymbol
        ALPHABET.isWildcard(firstSymbol)
            && ALPHABET.codeToWildcard(firstSymbol).matchesStrictly(secondSymbol) -> secondSymbol
        ALPHABET.isWildcard(secondSymbol)
            && ALPHABET.codeToWildcard(secondSymbol).matchesStrictly(firstSymbol) -> firstSymbol
        else -> ALPHABET.codeToWildcard(firstSymbol).union(ALPHABET.codeToWildcard(secondSymbol))
    }

    private fun Wildcard.union(second: Wildcard): Byte =
        ALPHABET.maskToWildcard(basicMask or second.basicMask).code

    private fun Wildcard.matchesStrictly(secondSymbol: Byte): Boolean = when {
        !ALPHABET.isWildcard(secondSymbol) -> matches(secondSymbol)
        else -> {
            val secondAsWildcard = ALPHABET.codeToWildcard(secondSymbol)
            basicMask xor secondAsWildcard.basicMask and secondAsWildcard.basicMask == 0L
        }
    }
}

fun Map<GeneFeature, CompositeMutations>.intersection(with: Map<GeneFeature, CompositeMutations>): Map<GeneFeature, CompositeMutations> =
    MutationsUtils.zip(this, with) { a, b, _ -> a.intersection(b) }

fun CompositeMutations.intersection(with: CompositeMutations): CompositeMutations =
    copy(mutationsFromParentToThis = mutationsFromParentToThis.intersection(with.mutationsFromParentToThis))
